lib/modules.nix

```diff
@@ -772,6 +772,23 @@ rec {
       else {};
   };
 
+  /* Evaluates to `value`, type-checking it lazily against `type`, and raising an
+     error if the type doesn't match.
+  */
+  typeChecking = { type, value, pos ? builtins.unsafeGetAttrPos "value" args, loc ? [ "value" ] }@args: let
+    merged = mergeDefinitions loc type [
+      { file = pos.file or unknownModule; inherit value; }
+    ];
+  in merged.mergedValue;
+
+  /* A simple deep type-checking function that returns a boolean, for use outside
+     of the module system.
+     Note that calling `type.check` is not enough: see `mkOptionType`'s documentation.
+  */
+  typeCheck = type: value: let
+    eval = builtins.tryEval (builtins.deepSeq (typeChecking { inherit type value; }) null);
```

Comment:
I guess it doesn't add much value, but you could simplify this to `typeChecking args`.

It improves `pos`, but that only shows up in a hypothetical `abort` message, but that should be a `throw` anyway. So the values is only in simplification and uniformity.

Reply:
I think I prefer the simplicity of `typeCheck t v`.

Reply:
Uniformity isn't as strong as interchangeability anyway, and we'll want to curry this one, so fair enough.